app/bundles/LeadBundle/Model/LeadModel.php

```diff
@@ -920,6 +920,75 @@ public function removeFromStages($lead, $stage, $manuallyRemoved = true)
         return $this;
     }
 
+    /**
+     * Add lead to Stage.
+     */
+    public function addToStage(Lead $lead, Stage $stage, string $origin): LeadModel
+    {
+        $lead->setStage($stage);
+        $lead->stageChangeLogEntry(
+            $stage,
+            $stage->getId().': '.$stage->getName(),
+            $origin
+        );
+
+        return $this;
+    }
+
+    public function changeStage(Lead $lead, Stage $stage, string $origin): void
+    {
+        // Get the current stage and validate it vs the new one
+        $currentStage = $lead->getStage();
+        if ($currentStage) {
+            if ($currentStage->getId() === $stage->getId()) {
+                throw new \UnexpectedValueException($this->translator->trans('mautic.stage.campaign.event.already_in_stage'));
+            }
+
+            if ($currentStage->getWeight() > $stage->getWeight()) {
+                throw new \UnexpectedValueException($this->translator->trans('mautic.stage.campaign.event.stage_invalid'));
+            }
+        }
+
+        $this->addToStage($lead, $stage, $origin);
+        $this->saveEntity($lead);
+
+        $this->logger->info(
+            sprintf(
+                'LeadBundle: Lead %s changed stage from %s (%s) to %s (%s) by %s',
+                $lead->getId(),
+                $currentStage?->getName(),
+                $currentStage?->getId(),
+                $stage->getName(),
+                $stage->getId(),
+                $origin
+            )
+        );
+    }
+
+    public function removeFromStage(Lead $lead, ?Stage $stage, string $origin): void
```

Comment:
Please, cover this method with a functional test.

I'm not a fan of the optional `$stage` param in the middle. Can a contact be removed from other stage than what is in the `Lead` entity? Perhaps it could be renamed to `removeFromCurrentStage(Lead $lead, string $origin)` with just 2 params?